config/store.go

```diff
@@ -188,124 +196,155 @@ func (s *Store) Set(newCfg *model.Config) (*model.Config, error) {
+	// We set the SiteURL to empty (if nil) so that the following call to
+	// SetDefaults() will generate missing data. This avoids an additional write
+	// to the backing store.
+	if loadedCfg.ServiceSettings.SiteURL == nil {
+		loadedCfg.ServiceSettings.SiteURL = model.NewString("")
+	}
 
-	loadedConfig = applyEnvironmentMap(loadedConfig, GetEnvironment())
+	// Setting defaults allows us to accept partial config objects.
+	loadedCfg.SetDefaults()
 
-	fixConfig(loadedConfig)
+	loadedCfgNoEnv := loadedCfg
+	fixConfig(loadedCfgNoEnv)
 
-	if err := loadedConfig.IsValid(); err != nil {
+	loadedCfg = applyEnvironmentMap(loadedCfg, GetEnvironment())
+	fixConfig(loadedCfg)
+	if err := loadedCfg.IsValid(); err != nil {
 		return errors.Wrap(err, "invalid config")
 	}
 
-	// Apply changes that may have happened on load to the backing store.
-	oldCfgBytes, err := json.Marshal(oldCfg)
-	if err != nil {
-		return errors.Wrap(err, "failed to marshal old config")
+	// Backing up feature flags section in case we need to restore them later on.
+	oldCfgFF := oldCfg.FeatureFlags
+	loadedCfgFF := loadedCfg.FeatureFlags
+	loadedCfgNoEnvFF := loadedCfgNoEnv.FeatureFlags
+	// Clearing FF sections to avoid both comparing and persisting them.
+	if s.readOnlyFF {
+		oldCfg.FeatureFlags = nil
+		loadedCfg.FeatureFlags = nil
+		loadedCfgNoEnv.FeatureFlags = nil
 	}
-	newCfgBytes, err := json.Marshal(loadedConfig)
+
+	// Check for changes that may have happened on load to the backing store.
+	hasChanged, err := confsDiff(oldCfg, loadedCfg)
 	if err != nil {
-		return errors.Wrap(err, "failed to marshal loaded config")
+		return errors.Wrap(err, "failed to compare configs")
 	}
 
-	var shouldStore bool
-	hasChanged := len(configBytes) == 0 || !bytes.Equal(oldCfgBytes, newCfgBytes)
-	if hasChanged {
-		featureFlags := s.configNoEnv.FeatureFlags
-		// Don't persist feature flags unless we are on MM cloud
-		// MM cloud uses config in the DB as a cache of the feature flag
-		// settings in case the management system is down when a pod starts.
-		if !s.persistFeatureFlags {
-			s.configNoEnv.FeatureFlags = loadedFeatureFlags
-		}
-		toStoreBytes, err := json.Marshal(s.configNoEnv)
-		if err != nil {
-			return errors.Wrap(err, "failed to marshal old config")
-		}
-		shouldStore = !bytes.Equal(toStoreBytes, configBytes)
-		// We write back to the backing store only if
-		// the config has changed and the store is not read-only.
-		if !s.readOnly && shouldStore {
-			err := s.backingStore.Set(s.configNoEnv)
-			s.configNoEnv.FeatureFlags = featureFlags
-			if err != nil && !errors.Is(err, ErrReadOnlyConfiguration) {
-				return errors.Wrap(err, "failed to persist")
-			}
+	// We write back to the backing store only if the store is not read-only
+	// and the config has either changed or is missing.
+	if !s.readOnly && (hasChanged || len(configBytes) == 0) {
+		err := s.backingStore.Set(loadedCfgNoEnv)
+		if err != nil && !errors.Is(err, ErrReadOnlyConfiguration) {
```

Comment:
Is there a need to check for `!errors.Is(err, ErrReadOnlyConfiguration)` here?

Reply:
This precedes me so I can only speculate but I believe we need that because internally the file won't get updated in case of a read-only config but we might still want to invoke the listeners and otherwise update the cached version.